### ISSUE TITLE
Allow make-style dependencies to be generated without fully compiling

### ISSUE THREAD

**Issue by BenHetherington** [rednex/rgbds#182]
At present, when using the `-M` flag, rgbasm fully compiles the given .asm file, and outputs a file containing its dependencies (e.g. due to `INCLUDE`s or `INCBIN`s). This is great for many scenarios involving Makefiles.

However, if a file isn't found, compilation stops. This is good for the proper compilation process, as the files must be openable to correctly generate the object file. However, assume a snippet like so:

```make
%.o: %.asm
%.o: %.asm $(DEPDIR)/%.d
	rgbasm -o '$@' -M $(DEPDIR)/$*.Td '$<'
	@mv $(DEPDIR)/$*.Td $(DEPDIR)/$*.d

%.2bpp : %.png
	rgbgfx -d 2 -o '$@' '$<'
```

As part of this, I convert PNGs to tile data using rgbgfx, and I'm treating the tile data as an intermediary. However, if I just add `INCBIN "Tiles.2bpp"` to an .asm file, and try the above dependency generation, it fails, because the .2bpp file doesn't yet exist. Instead, I'd have to manually add a `test.o : Tiles.2bpp` dependency to my Makefile.

To combat this, would it be possible to add a related (`-m`?) flag, which also generates dependencies in the same fashion, but doesn't (necessarily) compile the given file, and thus, not require that all files be present?

(I'm aware that GCC has the ability to generate dependencies without a full compile, but I've not tried it myself, so I don't know exactly how it behaves in comparison)

**Comment by AntonioND** [comment]
The problem is how to make the assembler to fail without printing a lot of messages related to missing definitions, for example... I don't think this should be the reason for a new flag, the old one should be enough, but the fix isn't trivial.

**Comment by ISSOtm** [comment]
Fixed by #428 (`-MG`).
That said, the pret folk have expressed that it'd be a slow process. What could be done to optimize build times with this?

**Comment by AntonioND** [comment]
Maybe having two passes of the assembler would help. In the first one only macros are expanded, includes are handled, etc... and that generates the dependencies.

**Comment by ISSOtm** [comment]
What was the reason for only having a single-pass assembler, again?

**Comment by mid-kid** [comment]
In the old two-pass assembler macros were re-expanded again, making assembly rather slow, and there wasn't really any reason to assemble in two passes when `rgblink` essentially serves as that second pass. Not to mention making the assembler single-pass simplified a lot of the logic.

I wouldn't make the assembler two-pass again, instead, disable whatever logic could be slow and redundant for the generation of dependencies.
That said, in my experience the expansion of the macros itself is rather slow (especially considering we use `const` a lot to simulate enums), and I'm not sure if anything other than code generation could be disabled/avoided to speed up the process.

The current issue in the pret projects is that all dependencies have to be known before anything can be built, so missing dependencies can be built in the correct order, and some files already take too noticeably long to assemble (order of 2-5 seconds depnding on machine/OS/optimization options), to double the time by using this feature on incremental builds. I don't consider macro-argument-based includes essential enough of a feature to sacrifice build speed, and listing all dependencies for conditional includes isn't particularly wrong, either. I understand why this feature is the "correct" way of doing things, and why people would use this (I would in my personal projects, as well), but it just doesn't scale very well.

**Comment by ISSOtm** [comment]
> That said, in my experience the expansion of the macros itself is rather slow (especially considering we use const a lot to simulate enums), and I'm not sure if anything other than code generation could be disabled/avoided to speed up the process.

If I remember correctly, the problem with macros is that the arguments can be expanded anywhere, and this means tokenization has to be redone on every call.
A possible solution would be introducing a type of macros with token arguments (more C-like, and perhaps ca65-like too), that would be more efficient. Of course, the downside then is having two types of macros.

Also, if the whole file wasn't reparsed up to the point the build failed, this could be a speed boost. But considering the complexity of implementing some kind of caching, I'm doubtful.

**Comment by ISSOtm** [comment]
Plus included file contents could be modified by the build/update process, so scratch that last idea.

**Comment by ISSOtm** [comment]
Implemented in 0.4.0 (`-MG`). While the process may be kind of slow, there's no way to speed it up without hacks of some sort, besides speeding up RGBASM itself.
Thus, closing as resolved; performance improvements are another problem...
